fs.go

```diff
@@ -196,121 +222,215 @@ func (tfs *tarfs) Sub(dir string) (fs.FS, error) {
 	return subfs, nil
 }
 
-func (tfs *tarfs) get(name, op string) (fs.DirEntry, error) {
-	if !fs.ValidPath(name) {
-		return nil, &fs.PathError{Op: op, Path: name, Err: fs.ErrInvalid}
+func (tfs *tarfs) get(op, path string) (entry, error) {
+	if !fs.ValidPath(path) {
+		return nil, newErr(op, path, fs.ErrInvalid)
 	}
 
-	e, ok := tfs.entries[name]
+	e, ok := tfs.entries[path]
 	if !ok {
-		return nil, &fs.PathError{Op: op, Path: name, Err: fs.ErrNotExist}
+		return nil, newErrNotExist(op, path)
 	}
 
-	return e, nil
+	return e.(entry), nil
 }
 
-type entries interface {
-	append(fs.DirEntry)
-	get() []fs.DirEntry
+type entry interface {
+	fs.DirEntry
+	size() int64
+	readdir(path string) ([]fs.DirEntry, error)
+	readfile(path string) ([]byte, error)
+	entries(op, path string) ([]fs.DirEntry, error)
+	open() (fs.File, error)
 }
 
-type fileInfo interface {
-	FileInfo() (fs.FileInfo, error)
+type regEntry struct {
+	fs.DirEntry
+	name   string
+	ra     io.ReaderAt
+	offset int64
 }
 
-type entry struct {
-	h       *tar.Header
-	b       []byte
-	entries []fs.DirEntry
-}
+var _ entry = &regEntry{}
 
-var _ fs.DirEntry = &entry{}
+func (e *regEntry) size() int64 {
+	info, _ := e.Info() // err is necessarily nil
+	return info.Size()
+}
 
-func (e *entry) Name() string {
-	return e.h.FileInfo().Name()
+func (e *regEntry) readdir(path string) ([]fs.DirEntry, error) {
+	return nil, newErrNotDir("readdir", path)
 }
 
-func (e *entry) IsDir() bool {
-	return e.h.FileInfo().IsDir()
+func (e *regEntry) readfile(path string) ([]byte, error) {
+	r, err := e.reader()
+	if err != nil {
+		return nil, err
+	}
+
+	b := bytes.NewBuffer(make([]byte, 0, e.size()))
+
+	if _, err := io.Copy(b, r); err != nil {
+		return nil, err
+	}
+
+	return b.Bytes(), nil
 }
 
-func (e *entry) Type() fs.FileMode {
-	return e.h.FileInfo().Mode() & fs.ModeType
+func (e *regEntry) entries(op, path string) ([]fs.DirEntry, error) {
+	return nil, newErrNotDir(op, path)
 }
 
-func (e *entry) Info() (fs.FileInfo, error) {
-	return e.h.FileInfo(), nil
+func (e *regEntry) open() (fs.File, error) {
+	r, err := e.reader()
+	if err != nil {
+		return nil, err
+	}
+
+	return &file{e, &readSeeker{&readCounter{r, 0}, e}, -1, false}, nil
 }
 
-var _ entries = &entry{}
+func (e *regEntry) reader() (io.Reader, error) {
+	tr := tar.NewReader(io.NewSectionReader(e.ra, e.offset, 1<<63-1))
```

Comment:
Found the 1.17 fix, I think:
```suggestion
	tr := tar.NewReader(io.NewSectionReader(e.ra, e.offset, 1<<63-1-e.offset))
```

See https://github.com/golang/go/issues/48620

Reply:
Good catch thanks for the fix